### DISCUSSION TITLE
Scalable Keyboard Simulation v2 (12.20.24)

### DISCUSSION THREAD

**Discussion by lageist** [UNC-Project-Open-AAC/OS-DPI#260] [Ideas]
The following represents a simulation user interface (UI) for a project aimed at designing a scalable keyboard that can be efficiently adjusted based on positioning and desired access method - specifically a large, eye gaze access-enabled keyboard filling the bottom two-thirds of the screen when in a seated position and a scaled down, trackball access-enabled comparable keyboard when lying down. The link below provides a simulation of a UI similar to the co-designer's current system (Text Talker from Smartbox) to explore the needed functionality. The scaled-up version in the current iteration enables head-tracking using the computer's built-in camera as an initial simulation. 

To explore the functionality, load the simulated UI using the link below: 
https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/user-attachments/files/18086076/Grid3_Patrick_12.10.24.zip

 Use the following keys on an external keyboard to explore the simulated functionality:
- **t** key to scale up and enable camera and head tracking
- **Shift** + **t** key to scale down
- **Alt** or **Option** key followed by the **d** key to toggle in/out of developer mode.

**Comment by KeyPatrick** [comment]
I was mistaken about the size of the font on the smaller keyboard.  It does not need to be that large, and it would be better if it was smaller because I can only see part of some of the buttons.

Also, re: the smaller keyboard.  I think that it would be ideal, if as Sophie suggested, it could be a pop up.  And if it could float so that I could move it around and also pin it that would be great.  But, if not, it would be better to have it positioned in the middle at the top, just under a message window.

**Reply by sofia-bg** [reply]
Hey Patrick - does this look more like what you are thinking for the smaller keyboard? I am not sure that we can emulate "moving it around" though @gbishop I suppose I could make tab controls with it in different places but not sure how Patrick would then "move" it. 

[Grid3_Patrick_1_2_25.zip](https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/user-attachments/files/18295750/Grid3_Patrick_1_2_25.zip)

**Comment by gbishop** [comment]
I bet we could do something with the ability to use state variables as
properties. I hallucinate that we could offer the ability to move the
keyboard around in small steps with button pushes (or whatever).

**Reply by gbishop** [reply]
Here is a quick hack using the < and > keys in the lower row on the tiny keyboard. I'm using the Excel-like feature that allows values of properties like "Scale" to be expressions by preceding the expression with an equal sign.

[Grid3_Patrick_1_2_25_lr.zip](https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/user-attachments/files/18300249/Grid3_Patrick_1_2_25_lr.zip)

**Reply by sofia-bg** [reply]
oh very cool!

**Comment by sofia-bg** [comment]
Hey @gbishop @KeyPatrick pointed out that dwell is still active on the small keyboard when the intention is to use the mouse to click. I tried using conditions to set the point and dwell to only work when $HeadMouse = "on" - that seems to have set it so that the dwell time remains on the large keyboard, but on the small now it dwells almost automatically. Any ideas? 

[Grid3_Patrick_1.6.25](https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/user-attachments/files/18325995/Grid3_Patrick_1_6_25.zip)

**Comment by gbishop** [comment]
Try this. I added the condition to every handler.
[Grid3_Patrick_1_8_25.zip](https://github.com/user-attachments/files/18349459/Grid3_Patrick_1_8_25.zip)

**Reply by gbishop** [reply]
Sorry for the delay. Our oven died in a very dangerous way and I've been removing it in preparation for a new one.

**Reply by sofia-bg** [reply]
Oh gosh that is scary! And thanks for the solution :)

**Reply by KeyPatrick** [reply]
Hopefully you were all safe Gary.  I will try out the change.  I haven't noticed any delay from you all.  This has been a very speedy way to work together and you are all making changes so fast, it is amazing.

**Comment by sofia-bg** [comment]
@gbishop already stuck with the modal - I created it but when I linked it to the gear button in actions it opens to a white page not the grid. Also, how can I change the relative position of the modal so that you can still see the keyboard? Seems like I need to see where the keyboard is moving in real time for it to be effective. 
[Grid3_Patrick_1_10_25_v2.zip](https://github.com/user-attachments/files/18381836/Grid3_Patrick_1_10_25_v2.zip)

**Comment by gbishop** [comment]
You are setting $tabControl to "move" but there is no tab panel with that
name...

Instead you should be opening the Modal by setting $modalOpen to 1.

**Reply by sofia-bg** [reply]
i have the modal opening and set the background to clear but the keyboard is still obscured by the grid, which i seem unable to change the transparency of. I tried adding gaps and setting them to be transparent but they just turned up white... any ideas? 

[Grid3_Patrick_1_17_25.zip](https://github.com/user-attachments/files/18458830/Grid3_Patrick_1_17_25.zip)

**Reply by gbishop** [reply]
This fixes it but I think it would be better with a row of buttons along the bottom, or something.
[Grid3_Patrick_1_17_25_gb.zip](https://github.com/user-attachments/files/18460724/Grid3_Patrick_1_17_25_gb.zip)

**Reply by sofia-bg** [reply]
i agree that a box grid probably doesn't make the most sense, but with patrick in mind using the track ball don't want it to be too far away that it takes a long way to scroll to... thanks will play more later

**Reply by gbishop** [reply]
In the future that Customize hack won't be required. I have changed the CSS so that div inherits its parent's background color.

**Comment by gbishop** [comment]
That's a bug! It is putting the color on the underlay instead of the modal.
I'll fix it.

**Comment by gbishop** [comment]
Maybe make another row appear below the keyboard? Then it is right there.

**Comment by sofia-bg** [comment]
I have two solutions here, the first was probably a bad idea using tab controls since it lets me move the keyboard within the tab that has the new row of buttons, but if i close and go back to the original tab it doesn't keep the placement. 

I have the modal working for left and right, but the gaps in the modal are still showing up white not transparent. I assume $lr stands for left right which is why I can't use the same solution for up down. If I try then it just also moves left or right. I tried putting in = 1.0 - $ud to create the state variable $ud but wondering if $lr is a state variable you "created" somewhere else? 

[new tab](https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/user-attachments/files/18512071/Grid3_Patrick_1_22_25.zip)

[modal](https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/user-attachments/files/18512304/Grid3_Patrick_1_17_25.zip)

**Reply by gbishop** [reply]
The version I posted here on the 25th supports up and down and is fully transparent. Look at it.

https://github.com/user-attachments/files/18460724/Grid3_Patrick_1_17_25_gb.zip

To make controls appear, I suggest you use a grid to represent the buttons in a single row and set its scale to 0 when you don't want to see it.

How about this?
[Grid3_Patrick_1_23_25.zip](https://github.com/user-attachments/files/18521746/Grid3_Patrick_1_23_25.zip)

**Reply by sofia-bg** [reply]
Oh that is an interesting solution, but I worry for Patrick using the track ball that they are now too far away. I can see how to move it up, and can play around with that. I also was able to get the modal to look how I want, even remove the border so you can't tell but then unlike your solution you can't use the keyboard while the modal is open. 

[Grid3_Patrick_modal](https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/user-attachments/files/18525397/Grid3_Patrick_1_23_25.zip)

**Reply by gbishop** [reply]
The problem with attaching the buttons to the keyboard is the buttons will move with the keyboard. So he'll have to use the trackball to reposition his cursor for every click.

**Reply by gbishop** [reply]
Check this version out. It might be the best of both worlds. I swapped the grid and the spacer on the bottom.
[Grid3_Patrick_1_23_25_a.zip](https://github.com/user-attachments/files/18527446/Grid3_Patrick_1_23_25_a.zip)
